__tests__/table_moveColumn.js

```diff
@@ -0,0 +1,476 @@
+            ),
+            row(td(p('2')), td({ rowspan: 2 }, p('---merged-row----'))),
+            row(td(p('4')), td(p('5')))
+          )
+        )
+      );
+
+      const newTr = moveColumn(0, 2)(tr);
+      expect(newTr.doc).toEqualDocument(
+        doc(
+          table(
+            row(
+              td(p('0')),
+              td(p('1')),
+              td({ rowspan: 2 }, p('---merged-row----'))
+            ),
+            row(td(p('2')), td({ rowspan: 2 }, p('---merged-row----'))),
+            row(td(p('5')), td(p('4')))
+          )
+        )
+      );
+    });
+  });
+
+  describe('on a complex table with merged cells and rows', () => {
+    it('keep the merged content columns order', () => {
+      const {
+        state: { tr }
+      } = createEditor(
+        doc(
+          table(
+            row(tdEmpty, td(p('1')), td(p('2'))),
+            row(tdEmpty, td({ colspan: 2 }, p('3'))),
+            row(tdEmpty, td(p('4')), td(p('5')))
+          )
+        )
+      );
+
+      const newTr = moveColumn(1, 0)(tr);
+      expect(newTr.doc).toEqualDocument(
+        doc(
+          table(
+            row(td(p('1')), td(p('2')), tdEmpty),
+            row(td({ colspan: 2 }, p('3')), tdEmpty),
+            row(td(p('4')), td(p('5')), tdEmpty)
+          )
+        )
+      );
+    });
+
+    it('should not move columns when there is not movement', () => {
+      const {
+        state: { tr }
+      } = createEditor(
+        doc(
+          table(
+            row(td({ colspan: 3 }, p('merged cell'))),
+            row(td(p('2')), tdCursor, tdEmpty),
+            row(td(p('3')), tdEmpty, tdEmpty)
+          )
+        )
+      );
+
+      const newTr = moveColumn(1, 0)(tr);
+      expect(newTr.doc).toEqualDocument(
+        doc(
+          table(
+            row(td({ colspan: 3 }, p('merged cell'))),
+            row(td(p('2')), tdCursor, tdEmpty),
+            row(td(p('3')), tdEmpty, tdEmpty)
+          )
+        )
+      );
+    });
+
+    describe('table 3x5', () => {
+      const {
+        state: { tr }
+      } = createEditor(
+        doc(
+          table(
+            row(
+              td({ rowspan: 2 }, p('M1')),
+              td(p('0')),
+              td(p('1')),
+              td(p('2')),
+              td(p('3'))
+            ),
+            row(
+              td({ colspan: 2 }, p('4')),
+              td({ rowspan: 2, colspan: 2 }, p('M2'))
+            ),
+            row(td(p('5')), td(p('6')), td(p('7')))
+          )
+        )
+      );
+
+      const expectedResult = doc(
+        table(
+          row(
+            td({ rowspan: 2 }, p('M1')),
+            td(p('2')),
+            td(p('3')),
+            td(p('0')),
+            td(p('1'))
+          ),
+          row(
+            td({ rowspan: 2, colspan: 2 }, p('M2')),
+            td({ colspan: 2 }, p('4'))
+          ),
+          row(td(p('5')), td(p('6')), td(p('7')))
+        )
+      );
+
+      it('should move column 1 to 3', () => {
```

Comment:
I think there is a misunderstand here, this is the original table:

<img width="858" alt="Screen Shot 2019-03-27 at 6 21 19 pm" src="https://user-images.githubusercontent.com/3452187/55057404-3b925e80-50bd-11e9-9f8a-b66a76758cec.png">

and this is the expected result:

<img width="845" alt="Screen Shot 2019-03-27 at 6 23 34 pm" src="https://user-images.githubusercontent.com/3452187/55057497-75fbfb80-50bd-11e9-9436-e1dac0c854e5.png">

Reply:
In this example we are moving column at index 1 to index 3 (the space between cells `a4` and `a5`), which should not be allowed because of the col-spanning cell `b2`. I am not sure if it's a good idea to make an assumption regarding where the moving columns should end up in case moving operation is not permitted at the destination index (shift the dropping columns to the left or to the right of the merged section). 
Thoughts @torifat ?

Reply:
I think we are confusing between UI and utils rules, for ProserMirror this is an array and we are not sending the element to be in the middle of something we are sending it to a specific position, like this:


```javascript
const arr = ['a', 'b', 'c', 'd', 'e'];

moveColumn(0, 3); // I'm not naming it "between cells 2-3" this is UI issue

// the result expected
['b', 'c', 'd', 'a', 'e']
````

How is going to decide where we should add the line between it's the UI that is covered on this repo.

Reply:
in your example:
```
const arr = ['a', 'b', 'c', 'd', 'e'];
```
imagine `d` and `e` items of the array cannot be logically split apart and it shouldn't be a valid operation regardless of the UI.

Reply:
I think both approaches are valid, we need to decide what path we want to. I'm keen to move with the current one because it's more practical if this lib solves this issue for us, without this we are going to add a lot of checks before calling this function, for nothing.

Reply:
And, we are not spliting the `d and e` we are moving them together.

Reply:
Ok, I'll throw a exception if you can't move to the target position, and I'll add two new parameters:

```
{
  tryToFit: boolean; // If true I'll run the current logic and try to fit
  direction: [-1, 1]; // A option to decide where I should put the position, if is null I'll decide using the origin and target direction
}
```